sound/soc/sof/intel/hda-dai.c

```diff
@@ -309,6 +324,18 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 			dev_err(sdev->dev, "%s: failed to clean up link DMA\n", __func__);
 			return ret;
 		}
+
+		/*
+		 * With IPC4 the pipelines (including the dai widget)) are not
+		 * freed up on stop/suspend.
+		 * If the stream is re-started after stop instead of closing it
+		 * then we need to keep the link DMA reserved.
+		 */
+		if (sdev->pdata->ipc_type == SOF_IPC_TYPE_4) {
```

Comment:
this likely need to be changed to:
```c
		if (cmd == SNDRV_PCM_TRIGGER_STOP &&
		    sdev->pdata->ipc_type == SOF_IPC_TYPE_4) {
```

Reply:
@ujfalusi why not just remove the call for cleanup then? What is the idea behind cleaning up the DMA and reclaiming it?

Reply:
yeah - should we not just give up the link DMA ID in close() ?

Reply:
Ahh I see the problem. Maybe we only need to clear the stream ID for stop without giving up the link DMA?

Reply:
@ujfalusi I assume your above description about how the link DMA is handled between driver/FW is commented inline ?

Reply:
@ujfalusi what I was asking was if it is enough if we do just 		snd_hdac_ext_bus_link_clear_stream_id(hlink, stream_tag); after clearing the RUN bit during stop?

Reply:
@ujfalusi do you think this will work?
```
diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index f5f2f832d117..3b7eb5736fcf 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -320,12 +320,26 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 
        switch (cmd) {
        case SNDRV_PCM_TRIGGER_STOP:
-               ret = hda_link_dma_cleanup(substream, hext_stream, dai);
-               if (ret < 0) {
-                       dev_err(sdev->dev, "%s: failed to clean up link DMA\n", __func__);
-                       return ret;
+       {
+               /* Clear the stream ID in the case of playback without giving up the link DMA ID */
+               if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+                       struct hdac_ext_stream *hext_stream;
+                       struct hdac_ext_link *hlink;
+                       int stream_tag;
+
+                       hlink = ops->get_hlink(sdev, substream);
+                       if (!hlink)
+                               return -EINVAL;
+
+                       hext_stream = ops->get_hext_stream(sdev, dai, substream);
+                       if (!hext_stream)
+                               return -EINVAL;
+
+                       stream_tag = hdac_stream(hext_stream)->stream_tag;
+                       snd_hdac_ext_bus_link_clear_stream_id(hlink, stream_tag);
                }
                break;
+       }
        case SNDRV_PCM_TRIGGER_SUSPEND:
        {
                struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
```

Reply:
BTW I tested this and it works for me on my HDA TGL machine

Reply:
@ranj063, this does not apply and it might re-introduce the metallic noise as we need to reset the link DMA, this patch will not do that, right?

What about this one:
```diff
diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index 2a6f821c9e14..1a6f13feef4c 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -103,8 +103,10 @@ hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai
 	return sdai->platform_private;
 }
 
-int hda_link_dma_cleanup(struct snd_pcm_substream *substream, struct hdac_ext_stream *hext_stream,
-			 struct snd_soc_dai *cpu_dai)
+static int
+hda_link_dma_cleanup(struct snd_pcm_substream *substream,
+		     struct hdac_ext_stream *hext_stream,
+		     struct snd_soc_dai *cpu_dai, bool release)
 {
 	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, cpu_dai);
 	struct sof_intel_hda_stream *hda_stream;
@@ -128,6 +130,12 @@ int hda_link_dma_cleanup(struct snd_pcm_substream *substream, struct hdac_ext_st
 		snd_hdac_ext_bus_link_clear_stream_id(hlink, stream_tag);
 	}
 
+	if (!release) {
+		/* force stream reconfiguration, including LinkDMA reset */
+		hext_stream->link_prepared = 0;
+		return 0;
+	}
+
 	if (ops->release_hext_stream)
 		ops->release_hext_stream(sdev, cpu_dai, substream);
 
@@ -211,7 +219,7 @@ static int __maybe_unused hda_dai_hw_free(struct snd_pcm_substream *substream,
 	if (!hext_stream)
 		return 0;
 
-	return hda_link_dma_cleanup(substream, hext_stream, cpu_dai);
+	return hda_link_dma_cleanup(substream, hext_stream, cpu_dai, true);
 }
 
 static int __maybe_unused hda_dai_hw_params_data(struct snd_pcm_substream *substream,
@@ -304,7 +312,8 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_STOP:
 	case SNDRV_PCM_TRIGGER_SUSPEND:
-		ret = hda_link_dma_cleanup(substream, hext_stream, dai);
+		ret = hda_link_dma_cleanup(substream, hext_stream, dai,
+					   cmd == SNDRV_PCM_TRIGGER_STOP ? false : true);
 		if (ret < 0) {
 			dev_err(sdev->dev, "%s: failed to clean up link DMA\n", __func__);
 			return ret;
@@ -673,7 +682,7 @@ static int hda_dai_suspend(struct hdac_bus *bus)
 
 			ret = hda_link_dma_cleanup(hext_stream->link_substream,
 						   hext_stream,
-						   cpu_dai);
+						   cpu_dai, true);
 			if (ret < 0)
 				return ret;
 		}
diff --git a/sound/soc/sof/intel/hda.h b/sound/soc/sof/intel/hda.h
index 22bd9c3c8216..ee4ccc1a5490 100644
--- a/sound/soc/sof/intel/hda.h
+++ b/sound/soc/sof/intel/hda.h
@@ -1038,8 +1038,6 @@ const struct hda_dai_widget_dma_ops *
 hda_select_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget);
 int hda_dai_config(struct snd_soc_dapm_widget *w, unsigned int flags,
 		   struct snd_sof_dai_config_data *data);
-int hda_link_dma_cleanup(struct snd_pcm_substream *substream, struct hdac_ext_stream *hext_stream,
-			 struct snd_soc_dai *cpu_dai);
 
 static inline struct snd_sof_dev *widget_to_sdev(struct snd_soc_dapm_widget *w)
 {
```

Reply:
Sorry @ujfalusi  my patch didn't apply as it was on top of my pr. But I love this version. Thanks!